core/raftlease/fsm.go

```diff
@@ -0,0 +1,411 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package raftlease
+
+import (
+	"io"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/raft"
+	"github.com/juju/errors"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/core/globalclock"
+	"github.com/juju/juju/core/lease"
+)
+
+const (
+	// CommandVersion is the current version of the command format. If
+	// this changes then we need to be sure that reading and applying
+	// commands for previous versions still works.
+	CommandVersion = 1
+
+	// SnapshotVersion is the current version of the snapshot
+	// format. Similarly, changes to the snapshot representation need
+	// to be backward-compatible.
+	SnapshotVersion = 1
+
+	// OperationClaim denotes claiming a new lease.
+	OperationClaim = "claim"
+
+	// OperationExtend denotes extending an already-held lease.
+	OperationExtend = "extend"
+
+	// OperationExpire denotes removing a lease after its duration.
+	OperationExpire = "expire"
+
+	// OperationSetTime denotes updating stored global time.
+	OperationSetTime = "setTime"
+)
+
+// NewFSM returns a new FSM to store lease information.
+func NewFSM() *FSM {
+	return &FSM{
+		entries: make(map[lease.Key]*entry),
+	}
+}
+
+// FSM stores the state of leases in the system.
+type FSM struct {
+	mu         sync.Mutex
+	globalTime time.Time
+	entries    map[lease.Key]*entry
+}
+
+func (f *FSM) claim(key lease.Key, holder string, duration time.Duration) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if _, found := f.entries[key]; found {
+		return lease.ErrInvalid
+	}
+	f.entries[key] = &entry{
+		holder:   holder,
+		start:    f.globalTime,
+		duration: duration,
+	}
+	return nil
+}
+
+func (f *FSM) extend(key lease.Key, holder string, duration time.Duration) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	entry, found := f.entries[key]
+	if !found {
+		return lease.ErrInvalid
+	}
+	if entry.holder != holder {
+		return lease.ErrInvalid
+	}
+	expiry := f.globalTime.Add(duration)
+	if !expiry.After(entry.start.Add(entry.duration)) {
+		// No extension needed - the lease already expires after the
+		// new time.
+		return nil
+	}
+	entry.start = f.globalTime
+	entry.duration = duration
+	return nil
+}
+
+func (f *FSM) expire(key lease.Key) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	entry, found := f.entries[key]
+	if !found {
+		return lease.ErrInvalid
+	}
+	expiry := entry.start.Add(entry.duration)
+	if !f.globalTime.After(expiry) {
+		return lease.ErrInvalid
+	}
+	delete(f.entries, key)
+	return nil
+}
+
+func (f *FSM) setTime(oldTime, newTime time.Time) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.globalTime != oldTime {
+		return globalclock.ErrConcurrentUpdate
+	}
+	f.globalTime = newTime
```

Comment:
Leases are stored as start time (when they were claimed/extended) + duration. (They could have been stored as just expiry times, but I think it was done like this to make the translation between global time and local time clearer - I copied this from the state lease implementation.) We could update all of the starts and durations, adding a second to the start and subtracting one from duration, whenever time moves forward, but there's no need to. In a non-distributed system where we were just using the wall clock for time (instead of a replicated time) we wouldn't update every lease each tick.

One related thing I did think about though, is that we could automatically remove leases that have expired here. The state-based lease store doesn't do this because it would require either triggers in mongo, or for the clock updater to know too much about the lease collection. But in the fsm they're right there, so it would be simple and potentially cut out a lot of raft traffic. I didn't want to make that change just yet because it's a bit subtle, but I think the lease manager would just handle it - it already needs to cope with a different manager expiring a lease out from under it.